File: Harvard_CS50_Introduction_to_CS/CS50-Pset6-master/vigenere.py
from sys import argv
from sys import exit
from cs50 import get_string


def main():
    # check the program running with one command-line argument
    if len(argv) != 2:
        print("Usage: python vigenere.py k")
        exit(1)
    # get ciper
    ciper = argv[1]
    # check if ciper is alphabets
    for item in ciper:
        if item.isalpha() == False:
            print("Usage: python vigenere.py k")
            exit(1)

    plaintext = get_string("plaintext: ")
    print("ciphertext: ", end="")
    pl = len(plaintext)
    cl = len(ciper)
    n = pl//cl

    # length of plaintext is less then that of ciper
    if pl <= cl:
        charCount = 0
        for i in range(0, pl):
            if plaintext[i].isalpha() == False:
                print(plaintext[i], end="")
            else:
                print(encrypt(plaintext[i], shift(ciper[charCount])), end="")
                charCount += 1
    else:
        charCount = 0
        # count entire plaintext'length to encrypt it's chars
        for i in range(0, pl):
            # check if the element is an alphabet, if not, simply print the element
            if plaintext[i].isalpha() == False:
                print(plaintext[i], end="")
            # if yes, print the encrypted char
            else:
                # count from 0 to pl/cl, (pl/cl-1)*cl + (l - 1) = pl - 1.
                # so maximum count is length of plaintext - 1
                # from that we can then substract the space and non-alphabet elements
                # anh the round of cipher is reversed.
                for j in range(0, n):
                    for l in range(0, cl):
                        if charCount == (j * cl + l):
                            key = shift(ciper[l])
                            print(encrypt(plaintext[i], key), end="")
                            break
                charCount += 1
    print()
    exit(0)


def shift(c):
    if c.isupper():
        return (ord(c) - 65)
    else:
        return (ord(c) - 97)


def encrypt(char, key):
    if char.isupper():
        return chr((ord(char) - 65 + key) % 26 + 65)
    elif char.islower():
        return chr((ord(char) - 97 + key) % 26 + 97)
    else:
        return char


if __name__ == '__main__':
    main()

File: Harvard_CS50_Introduction_to_CS/CS50-Pset2-master/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

int main(int argc, string argv[])
{
    //check the program running with one command-line argument
    if (argc != 2)
    {
        printf("Usage: ./caesar key\n");
        return 0;
    }
    char *cipers = argv[1];
    int length = strlen(cipers);
    int shift(char c);

    //if cipers are characters
    bool checkChar(char c);
    for (int i = 0; i < length; i++)
    {
        if (checkChar(cipers[i]))
        {
            continue;
        }
        else
        {
            printf("Usage: ./caesar key\n");
            return 0;
        }
    }
    char *plaintext = get_string("plaintext: ");
    int n = strlen(plaintext);
    int x = n/length;
    printf("cipertext: ");
    char encrypt(char c, int key);
    // if plaintext's length is smaller than and equal the ciper's length
    // then check if characters in ciper are alphabets
    // if not, print the character
    // if yes, encrypt the character by key
    if (n <= length)
    {
        int charCount = 0;
        for (int l = 0; l < n; l++)
        {
            if (!checkChar(plaintext[l]))
            {
                printf("%c", plaintext[l]);
            }
            else
            {
                printf("%c", encrypt(plaintext[l], shift(cipers[charCount])));
                charCount += 1;
            }

        }
    }
    //if plaintext's length is greater than the ciper's length
    // then check if characters in ciper are alphabets
    // if not, print the character
    // if yes, encrypt the character by key
    else
    {
        int charCount = 0;
        for (int l = 0; l < n; l++)
        {
            if (!checkChar(plaintext[l]))
            {
                printf("%c", plaintext[l]);
            }
            else
            {
                for (int i = 0; i < x; i++)
                {
                    for (int j = 0; j < length; j++)
                    {
                        if (charCount == (i*length + j)) //use charCount to get j
                        {
                            printf("%c", encrypt(plaintext[l], shift(cipers[j])));

                            break;
                        }
                    }
                }
                charCount = charCount + 1;
            }
        }
    }
    printf("\n");
    return 0;
}

//shift value
int shift(char c)
{
    if (isupper(c))
    {
        return (c - 65);
    }
    else if (islower(c))
    {
        return (c - 97);
    }
    else
    {
        return c;
    }
}
//check lower or upper case chars
char encrypt(char c, int key)
{
    if (isupper(c))
    {
        return ((c - 65 + key)%26 + 65);
    }
    else if (islower(c))
    {
        return ((c - 97 + key)%26 + 97);
    }
    else
    {
        return c;
    }
}
//check if the input is either character or not
bool checkChar(char c)
{
    if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'))
    {
        return true;
    }
    else
    {
        return false;
    }
}